### PR TITLE
feat: add v2.2.0 sync infrastructure, CI/CD, and comprehensive tests

### PR DESCRIPTION
**Pull request by hyunhee-jo**
## Objective

When opendataloader-pdf releases a new version (e.g., v2.1.0 adding `detect_strikethrough`), there is no automated way to sync the new options into the langchain wrapper. Every update requires manual parameter copy-paste, README edits, and ad-hoc testing — error-prone and slow. The project also lacks CI to catch regressions on PRs.

## Approach

Build three layers of automation — sync, test, release — so upstream releases flow into the wrapper with minimal manual work.

**1. Sync markers + upstream workflow** (`sync-upstream.yml`)
Insert `BEGIN/END SYNCED PARAMS` markers in `document_loaders.py` and README so the upstream code generator (`generate-langchain.mjs`) can patch parameters without touching surrounding code. The workflow triggers on `repository_dispatch` or manual input, generates code, runs tests, and creates a PR (draft on test failure, ready on success). Breaking changes (removed params) are auto-detected and labeled.

**2. PR + full test workflows** (`test.yml`, `test-full.yml`)
Every PR runs unit tests across Python 3.10/3.12/3.13 with `--disable-socket` network isolation, plus a minimum dependency test (langchain-core 1.0.0 + opendataloader-pdf 2.1.0). Manual full tests cover 3 OS × 2 Python with Java integration.

**3. Comprehensive test expansion** (34 new tests → 117 total)
Unit: Path objects, `detect_strikethrough`, JSON edge cases, format validation, temp cleanup.
Integration: `replace_invalid_chars`, `content_safety_off`, image options, directory loading.
Regression: snapshot comparison (text/markdown/json), metadata consistency, instance isolation.

**Security hardening:**
- All `run:` blocks use shell `$VERSION` instead of `${{ env.VERSION }}` to prevent injection from `repository_dispatch` payloads
- Cross-job outputs (`needs.*`) mapped to env vars before shell evaluation
- Top-level `permissions: {}` with per-job overrides
- Slack webhook guarded via `env` intermediary (secrets not accessible in `if:` conditions)

**Additional changes:**
- Add `detect_strikethrough` parameter (new in opendataloader-pdf v2.1.0)
- Bump minimum dependency to `opendataloader-pdf>=2.1.0`
- Fix `hybrid_timeout` docstring default: `"30000"` → `"0"`
- Fix README AI-AGENT-SUMMARY license: MIT → Apache-2.0
- Add CHANGELOG.md and pre-release checklist design doc

## Evidence

Loaded a real PDF (`samples/pdf/lorem.pdf`) with the updated wrapper under user-identical conditions:

| Scenario | Expected | Actual |
|----------|----------|--------|
| `detect_strikethrough=True`, format=markdown | Load succeeds, no error | 1 doc, 462 chars — OK |
| `detect_strikethrough` omitted (default) | Load succeeds, backward compatible | 1 doc, 462 chars — OK |
| `split_pages=True` (keyword arg, moved position) | Per-page documents | 1 page doc — OK |
| `split_pages=False` (keyword arg) | Single whole document | 1 whole doc — OK |
| `format="xml"` (invalid) | `ValueError` raised | `ValueError: Invalid format 'xml'. Valid options are: 'json', 'text', 'html', 'markdown'` |
| opendataloader-pdf 2.2.0 `convert()` signature | `detect_strikethrough` param exists | `True` |

Cannot verify `sync-upstream.yml` end-to-end — requires an actual upstream release to trigger `repository_dispatch`. Workflow syntax and job outputs validated via CI dry-run.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/workflows/sync-upstream.yml [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+          if [ "${{ needs.generate.outputs.breaking }}" == "true" ]; then
+            echo "" >> /tmp/pr-body.md
+            echo "### ⚠️ Breaking Changes Detected" >> /tmp/pr-body.md
+            cat .sync-report/breaking-changes.txt >> /tmp/pr-body.md 2>/dev/null || echo "See diff for details" >> /tmp/pr-body.md
+          fi
+
+          echo "" >> /tmp/pr-body.md
+          echo "### Changes" >> /tmp/pr-body.md
+          echo '```' >> /tmp/pr-body.md
+          git diff HEAD~1 --stat >> /tmp/pr-body.md
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml]: Fixed — added `fetch-depth: 2` to create-pr checkout.

File: .github/workflows/sync-upstream.yml [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        run: |
+          BRANCH="sync/v${{ env.VERSION }}"
+          echo "name=$BRANCH" >> $GITHUB_OUTPUT
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+          git checkout -b "$BRANCH"
+          git add langchain_opendataloader_pdf/document_loaders.py pyproject.toml README.md
+          git add .sync-report/ || true
+          git commit -m "sync: update to opendataloader-pdf v${{ env.VERSION }}"
+          git push origin "$BRANCH"
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml]: Fixed — added `--force-with-lease` to git push for idempotent reruns.

File: .github/workflows/sync-upstream.yml [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            echo "Attempt $i/20: waiting 30s..."
+            sleep 30
+          done
+          echo "::error::PyPI version ${{ env.VERSION }} not available after 10 minutes"
+          exit 1
+
+      - name: Install dependencies
+        run: |
+          pip install -e . pytest pytest-socket
+          pip install "opendataloader-pdf>=${{ env.VERSION }}"
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml]: Fixed — changed to `always() && needs.generate.result == 'success'` so create-pr runs even when test fails (draft PR).

File: tests/test_regression.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        loader = OpenDataLoaderPDFLoader(
+            file_path=str(LOREM_PDF), format="text", quiet=True, split_pages=False
+        )
+        docs = loader.load()
+        assert len(docs) == 1
+        actual = docs[0].page_content.strip()
+
+        if not self.SNAPSHOT_FILE.exists():
+            self.SNAPSHOT_FILE.write_text(actual, encoding="utf-8")
+            pytest.skip("Snapshot created, run again to verify")
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply tests/test_regression.py]: Fixed — snapshots now fail if missing in CI. Use `UPDATE_SNAPSHOTS=1` to opt-in creation locally.

File: langchain_opendataloader_pdf/document_loaders.py [hunk 5]
```diff
@@ -61,12 +62,14 @@ def __init__(
         sanitize: bool = False,
         pages: Optional[str] = None,
         include_header_footer: bool = False,
-        split_pages: bool = True,
+        detect_strikethrough: bool = False,
         hybrid: Optional[str] = None,
         hybrid_mode: Optional[str] = None,
         hybrid_url: Optional[str] = None,
         hybrid_timeout: Optional[str] = None,
         hybrid_fallback: bool = False,
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:70]: Fixed — `hybrid_timeout` docstring updated to `"0" (no timeout)`. Also bumped min dep to `>=2.1.0`.

File: langchain_opendataloader_pdf/document_loaders.py [hunk 6]
```diff
@@ -61,12 +62,14 @@ def __init__(
...
+        # --- END SYNCED PARAMS ---
+        split_pages: bool = True,
     ):
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:73]: Acknowledged — `split_pages` position changed but all documented examples use keyword args. Positional usage is not part of the public API contract.

File: langchain_opendataloader_pdf/document_loaders.py [hunk 7]
```diff
@@ -256,20 +263,22 @@ def lazy_load(self) -> Iterator[Document]:
...
-                markdown_page_separator=page_sep,
-                text_page_separator=page_sep,
-                html_page_separator=page_sep,
                 image_output=self.image_output,
                 image_format=self.image_format,
                 image_dir=self.image_dir,
                 sanitize=self.sanitize,
                 pages=self.pages,
                 include_header_footer=self.include_header_footer,
+                detect_strikethrough=self.detect_strikethrough,
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:272]: Fixed — bumped minimum dependency to `opendataloader-pdf>=2.1.0` (where `detect_strikethrough` was added). Min-dep test updated accordingly.

File: .github/workflows/sync-upstream.yml [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    timeout-minutes: 5
+    permissions:
+      contents: write
+    outputs:
+      changed: ${{ steps.changes.outputs.changed }}
+      breaking: ${{ steps.breaking.outputs.detected }}
+      branch: ${{ steps.branch.outputs.name }}
+
+    env:
+      VERSION: ${{ github.event.client_payload.version || github.event.inputs.version }}
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml:33]: Fixed — added semver validation step that rejects non-matching input.

File: .github/workflows/sync-upstream.yml [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+      - name: Run unit tests
+        run: pytest tests/test_document_loaders.py -v --disable-socket --allow-unix-socket
+
+  # =================================================================
+  # Job 3: Create PR + notify
+  # =================================================================
+  create-pr:
+    needs: [generate, test]
+    if: needs.generate.result == 'success' && needs.generate.outputs.changed == 'true'
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml]: Fixed — `create-pr` now uses `always() && needs.generate.result == 'success'` to handle test failures with draft PR.

File: .github/workflows/sync-upstream.yml [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+      - name: Determine PR status
+        id: status
+        run: |
+          if [ "${{ needs.test.result }}" == "success" ]; then
+            echo "draft=false" >> $GITHUB_OUTPUT
+            echo "result=✅ All tests passed" >> $GITHUB_OUTPUT
+          else
+            echo "draft=true" >> $GITHUB_OUTPUT
+            echo "result=❌ Tests failed" >> $GITHUB_OUTPUT
+          fi
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml:210]: Acknowledged — the `always()` condition now handles this case. Test skip → draft PR.

File: docs/plans/2026-04-02-pre-release-checklist-design.md [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.10'
+          cache: pip
+      - name: Install minimum dependency versions
+        run: |
+          pip install "langchain-core==1.0.0" "opendataloader-pdf==2.0.0" pytest
+          pip install -e . --no-deps
+      - run: pytest tests/test_document_loaders.py -v --disable-socket --allow-unix-socket
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply docs/plans/2026-04-02-pre-release-checklist-design.md]: Noted — internal design doc. Will update when finalizing.

File: tests/test_regression.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+        docs1 = loader1.load()
+        docs2 = loader2.load()
+
+        assert len(docs1) >= 1
+        assert len(docs2) >= 1
+        assert docs1[0].metadata["format"] == "text"
+        assert docs2[0].metadata["format"] == "markdown"
+        # Content should differ (different formats)
+        assert docs1[0].page_content != docs2[0].page_content
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply tests/test_regression.py:166]: Fixed — renamed to `test_sequential_loaders_no_conflict`.

File: langchain_opendataloader_pdf/document_loaders.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:272]: Fixed — bumped minimum dependency to `opendataloader-pdf>=2.1.0` where `detect_strikethrough` was added. Min-dep test updated accordingly.

[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:272]: Fixed — bumped minimum dependency to `opendataloader-pdf>=2.1.0` where `detect_strikethrough` was added. Min-dep test updated accordingly.

File: .github/workflows/sync-upstream.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml:33]: Fixed — added semver validation step (`^[0-9]+\.[0-9]+\.[0-9]+$`) to reject malformed version input before any processing.

File: .github/workflows/sync-upstream.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml:210]: Fixed — `create-pr` now uses `always()` so it runs even when test is skipped/failed, creating a draft PR.

File: langchain_opendataloader_pdf/document_loaders.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply langchain_opendataloader_pdf/document_loaders.py:70]: Fixed — docstring updated to `"0" (no timeout)`. Minimum dep bumped to `>=2.1.0`.

File: tests/test_regression.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply tests/test_regression.py:166]: Fixed — renamed to `test_sequential_loaders_no_conflict`.

File: tests/test_regression.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply tests/test_regression.py]: Fixed — snapshots now fail if missing in CI. Use `UPDATE_SNAPSHOTS=1` to opt-in creation locally.

File: .github/workflows/sync-upstream.yml [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+          LABELS="sync,automated"
+          [ "${{ needs.generate.outputs.breaking }}" == "true" ] && LABELS="$LABELS,breaking-change"
+
+          gh pr create \
+            --title "sync: update to opendataloader-pdf v${{ env.VERSION }}" \
+            --body-file /tmp/pr-body.md \
+            --base main \
+            $DRAFT_FLAG \
+            --label "$LABELS"
```
[the remark this answers is not in the document]
Comment (hyunhee-jo) [reply .github/workflows/sync-upstream.yml]: Fixed — added `fetch-depth: 2` to create-pr checkout.

**Review by bundolee — CHANGES_REQUESTED** [review]
## Review by benedict

전체적으로 잘 만들어진 PR입니다. Sync marker 설계, CI/CD 보안 구성, 테스트 커버리지 확장(83→117) 모두 좋습니다. 아래 3가지 수정 후 머지하겠습니다.

### Must Fix

**1. `sync-upstream.yml` shell injection 위험**

`${{ env.VERSION }}`은 GitHub Actions 템플릿 엔진이 bash 실행 전에 확장하므로, `repository_dispatch` payload에 shell 메타문자가 포함되면 validation 단계에서 이미 injection이 발생합니다. 모든 `run:` 블록에서 `${{ env.VERSION }}` 대신 `$VERSION` (shell 변수)을 사용해주세요.

```yaml
# Before (injectable)
if ! echo "${{ env.VERSION }}" | grep -qE ...
# After (safe)
if ! echo "$VERSION" | grep -qE ...
```

**2. JSON snapshot 테스트에 `UPDATE_SNAPSHOTS` 가드 누락**

`test_json_structure_matches_snapshot`에서 snapshot이 없을 때 무조건 생성하고 skip합니다. text/markdown 테스트처럼 `UPDATE_SNAPSHOTS` 체크를 추가해주세요. CI에서 snapshot 없이 실행하면 silent skip이 됩니다.

**3. PR body heredoc 들여쓰기 문제**

`Build PR body` 스텝의 heredoc 내용이 YAML 블록과 함께 들여쓰기되어 있어서, `##` 헤딩에 앞공백이 들어가 preformatted text로 렌더링됩니다. heredoc 내용을 dedent하거나 `<<-`를 사용해주세요.

### Suggestions

- `create-pr` job의 `git diff HEAD~1`은 merge commit 시 fragile합니다. `git diff main...HEAD --stat` 고려해주세요.
- design doc의 min-dep test가 `opendataloader-pdf==2.0.0`으로 되어있는데 실제 구현은 `2.1.0`입니다. 문서 업데이트 부탁합니다.
- README params table 기본값 quoting이 불일치합니다 (일부 quoted, 일부 unquoted). 자동 생성 기반이니 통일하면 좋겠습니다.

**Review by bundolee — CHANGES_REQUESTED** [review]
## Review by benedict

Well-structured PR overall. Sync marker design, CI/CD security posture, and test coverage expansion (83→117) are all solid. Please fix the 3 items below, then I'll merge.

### Must Fix

**1. Shell injection risk in `sync-upstream.yml`**

`${{ env.VERSION }}` is expanded by the GitHub Actions template engine before bash runs, so if a `repository_dispatch` payload contains shell metacharacters, injection occurs *during* the validation step itself. Use `$VERSION` (shell variable) instead of `${{ env.VERSION }}` in all `run:` blocks.

```yaml
# Before (injectable)
if ! echo "${{ env.VERSION }}" | grep -qE ...
# After (safe)
if ! echo "$VERSION" | grep -qE ...
```

**2. Missing `UPDATE_SNAPSHOTS` guard on JSON snapshot test**

`test_json_structure_matches_snapshot` creates the snapshot unconditionally when missing and skips. The text/markdown tests correctly gate on `UPDATE_SNAPSHOTS` — apply the same pattern here. Without it, CI silently skips instead of failing when the snapshot is absent.

**3. PR body heredoc indentation issue**

The heredoc content in the `Build PR body` step is indented along with the YAML block, so `##` headings will have leading whitespace and render as preformatted text. Dedent the heredoc content or use `<<-` with tabs.

### Suggestions

- `git diff HEAD~1` in the `create-pr` job is fragile if the branch has merge commits. Consider `git diff main...HEAD --stat` instead.
- The design doc still references `opendataloader-pdf==2.0.0` for the min-dep test, but the actual implementation correctly uses `2.1.0`. Worth updating for consistency.
- README params table has inconsistent quoting of defaults (some quoted, some not). Since this will be auto-generated going forward, worth normalizing.

**Review by bundolee — APPROVED** [review]
All 3 must-fix items confirmed resolved. LGTM — merging.
